File: Self/kmp.c
/*
*在t字符串中查找p子串
*/
int KMP (char * t,char * p,int *next)
{
    int i =0 ,j = 0;

    while (i < strlen(t) && j < strlen(p))
    {
        if(j = -1 || t[i] == p[j])
        {
            i++;
            j++;
        }
        else
        {
            //next数组根据PMT（部分匹配表）计算得出
            j = next[j];
        }
    }

    //p长度完全匹配
    if(j == strlen(p))
    {
        return i - j;
    }
    else
    {
        return -1;
    }
}

int * getNext(char *p)
{
    static next[100] = {0};
    int i = 0,j = -1;
    //便于程序计算将next[0]直接赋值为-1
    next[0] = -1;

    while (i < strlen(p))
    {
        if(j == -1 || p[i] == p[j])
        {
            ++i;++j;
            next[i] = j;
        }
        else
        {
            j = next[j];
        }
    }
}

File: Self/kmp.py
'''kmp'''
def kmp(source = "", find = ""):
    PMT = [0 for x in range(0,len(find))]
    #先计算pmt数组
    PMT[0] = -1
    i,j = 0,-1
    while i < len(find) - 1:
        if j == -1 or find[i] == find[j]:
            i = i + 1
            j = j + 1
            PMT[i] = j
        else:
            j = PMT[j]

    i ,j = 0 , 0
    while i < len(source) and j < len(find):
        if j == -1 or source[i] == find[j]:
            i = i + 1
            j = j + 1
        else:
            j = PMT[j]
     
    if j == len(find):
        return i - j
    else:
        return -1

if __name__ == "__main__":
    print("haha")
    str1 = "abababcaa"
    str2 = "ababababcasacbababasbcaabababavbcaababaebabcaababababcaababababca"
    print(kmp(str2,str1))
